Add `yq` subcommands to `limactl` and `lima-guestagent`

jandubois opened the pull request:
Adding it to `lime-guestagent` allows us to use `yq` in provisioning scripts without having to install it from a package repository or downloading it from GitHub.

Adding it to `limactl` makes it available to integration tests without having to install it. It` is essentially free because we already use `yqlib` inside `limactl` anyways, so the executable size doesn't really change.

The size of the (compressed) guestagent for aarch64 increases from 12MB to 14MB, which seems acceptable.

These are "hidden" commands that don't show up in --help output.

```console
❯ limactl yq --version
yq (https://github.com/mikefarah/yq/) version v4.47.1

❯ limactl yq -n '.foo="bar"'
foo: bar

❯ lima lima-guestagent yq -n '.foo="bar"'
foo: bar
```

This PR is meant to support the "yq provisioning mode" proposed in #3892, but I think is also valuable on its own.

---

I was initially thinking about making this a busybox-style "multi-call" binary, so you could put a symlink called `yq` into `/usr/local/bin` and have it work as a regular `yq` binary. But I'm worried about shadowing a `yq` version installed by the user with a shadow manager, so didn't add this functionality.

jandubois commented:
We probably should include a comment that the `yq` command implementation is copied from the upstream repo, and maybe also include the license file in our repo.

jandubois commented on cmd/lima-guestagent/main_linux.go:
```diff
@@ -14,6 +16,25 @@ import (
 
 func main() {
+	// `lima-guestagent yq` executes the embedded `yq` command instead.
+	if len(os.Args) > 1 && os.Args[1] == "yq" {
```
Because I had bad experiences with it before. I can't remember the details, but global options defined by one app could not be removed when calling a different app.

Maybe this was due to the complexity of the apps involved (I was also embedding `kubectl`), but I wasted a lot of time on it.

I also don't really see the point (what problem would that be solving?), given that this is essentially a multi-call binary setup where the embedded binary is not related to the rest of the options. Right now the code is directly lifted from the `yq` `main` package (because you cannot import it). Why would we want to complicate it?

If we wanted to show the subcommand in the `--help` output, then we could still add it to Cobra, but I would still want to delegate the `main` functionality before making the first Cobra call.

jandubois replied:
I've moved out the code copied from `yq` into its own function to make the separation clearer.

AkihiroSuda commented on LICENSE_YQ:
```diff
@@ -0,0 +1,30 @@
+Below is a copy of https://github.com/mikefarah/yq/blob/master/LICENSE
+
+This file is included in the Lima repository because `limactl yq` and
```
Why do we need this file specifically for yq?
We do not copy the license files for other Go dependencies

jandubois replied:
I'm not sure we need it.

I felt this is different because we are effectively bundling/distributing the `yq` command now, both for the host and the VM.

It is just compiled into our binary, which is just a packaging decision (share statically linked libraries). But otherwise the original utility is fully exposed to the end user, which is conceptually different from just using it as a library:

```console
❯ limactl yq --version
yq (https://github.com/mikefarah/yq/) version v4.47.1
```

I will remove it if you think we don't need it.

AkihiroSuda replied:
The license file is included in https://github.com/lima-vm/lima/releases/download/v1.2.1/lima-1.2.1-go-mod-vendor.tar.gz

(If this is not enough, we have to duplicate the license files for all the deps, not just for yq)

jandubois replied:
Good enough for me, I will remove it.

AkihiroSuda commented on cmd/limactl/main.go:
```diff
@@ -31,7 +32,34 @@ const (
 	advancedCommand     = "advanced"
 )
 
+// yqmain is a copy of main() in https://github.com/mikefarah/yq/blob/master/yq.go
```
Better to split a package, with an appropriate SPDX headers

jandubois replied:
What is the appropriate license header? I think it needs to maintain the MIT license and include the original copyright in addition to ours.

AkihiroSuda commented on cmd/yq/yq.go:
```diff
@@ -0,0 +1,45 @@
+// SPDX-FileCopyrightText: Copyright (c) 2017 Mike Farah
+// SPDX-License-Identifier: MIT
+
+// This file has been adapted from https://github.com/mikefarah/yq/blob/master/yq.go
```
s/master/<actual version you referred>/

jandubois commented:
I've added a second commit to add a `yq.lima` symlink both on the host and inside the VM (assuming the guestagent install directory is on the PATH; at least for provisioning scripts I've added it explicitly).

I thought this fits in with our other `docker.lima`, `kubectl.lima` custom versions.

I've kept it in a separate commit, in case we decide we don't want to do this. I can squash if we want to keep.

I also have not tested the Windows command that copies the file instead of making a symlink.

jandubois commented:
I'm kind of torn about installing `yq.lima` on Windows. I want to do it for consistency, but it feels wasteful to store a copy of `limactl` just because you don't have symlinks when you can also just use `limactl yq` instead of `yq.lima`.

I think I prefer to not install it on Windows; what do you think?

jandubois commented:
> I think I prefer to not install it on Windows; what do you think?

An alternative is to just create a symlink on Windows too, and ignore the failure if the user doesn't have "Developer Mode" enabled on their system (to allow non-admins to create symlinks).

jandubois commented:
> The convention of `<SOMETHING>.lima` is to run `<SOMETHING>` inside an instance of Lima.

Not really. It is a variant of the tool installed by Lima, that will automatically work with the daemons inside a Lima instance. But the tools run locally, as long as they are installed, and only fall back to running inside the VM when they don't exist locally.

The only tools that always run inside the VM are `apptainer.lima` and `nerdctl.lima`. All of `docker.lima`, `kubectl.lima`, and `podman.lima` run on the host.

The goal is to make the difference invisible. The `.lima` suffix is used to avoid shadowing the native tool.

afbjorklund commented:
> An alternative is to just create a symlink on Windows too

Couldn't you use a shell wrapper on Windows, and call `limactl yq`

> All of `docker.lima`, `kubectl.lima`, and `podman.lima` run on the host.

The main difference is that they just call the external tool, like `yq` here

afbjorklund commented on Makefile:
```diff
@@ -277,10 +277,17 @@ LIMA_CMDS = $(sort lima lima$(bat)) # $(sort ...) deduplicates the list
+ifneq ($(GOOS),windows)
+	ln -sf limactl$(exe) $@
+else
+	cp -aL $< $@
```
You could copy a cmd/yq.lima script here instead of making a copy

afbjorklund commented on Makefile:
```diff
@@ -493,6 +500,7 @@ uninstall:
 		"$(DEST)/bin/docker.lima" \
 		"$(DEST)/bin/podman.lima" \
 		"$(DEST)/bin/kubectl.lima" \
+		"$(DEST)/bin/yq.lima$(exe)" \
```
The exe would have to be removed here, as well (if it is a script)

AkihiroSuda reviewed (changes requested):
I still don't think we should expose this dev tool to end users in the same way as other `*.lima` helpers

jandubois commented:
> I still don't think we should expose this dev tool to end users in the same way as other `*.lima` helpers

I don't know, I thought it was useful, and think it fits the scheme of adding useful utilities with a `.lima` suffix. Especially `lima-hostagent yq` is a bit of a mouthful, but then it will probably be hidden by the `provision.mode: yq` feature anyways.

But whatever, I don't care that much, and have removed the extra commit. I only kept the change that `yq.lima.exe` would invoke `yq`, if the user makes the symlink themselves.
